### ISSUE TITLE
Reduce code duplication in interpolate and make it more generic

### ISSUE THREAD

**Issue by fmassa** [pytorch/pytorch#10482]
Currently, `torch.nn.functional.interpolate` has dedicated implementation for 1d, 2d and 3d data, as well as for nearest, bilinear and bicubic interpolation. The set of different kernels that are dispatched can be seen [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2050-L2079).

Those implementation are all very similar one to the other, and there is a lot of code duplication in there.
Compare for example [nearest](https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/SpatialUpSamplingNearest.c#L31-L95) with [bilinear](https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/SpatialUpSamplingBilinear.c#L33-L106).

I believe it is possible to refactor the underlying implementation so that we can have a single C++ / CUDA codepath, with minimal  code duplication.

This could be achieved in two independent steps (that could be done at the same time):
- factor out the different kernel computations, so that we have a generic `filter` that is used to compute the interpolation, plus the size of the filter as a struct. For an example, [see how Pillow implements it](https://github.com/python-pillow/Pillow/blob/master/src/libImaging/Resample.c#L9-L83), and then the interpolation coefficients can be generically computed [as in here](https://github.com/python-pillow/Pillow/blob/master/src/libImaging/Resample.c#L236).
- make the interpolate kernel separable (it is already separable in most cases). This means that we can have the user specify the dimensions he wants to interpolate as a list, and in the C++/CUDA code we have a loop over the dimensions. Something like `F.interpolate(image, dim=[-2, -1])` for spatial interpolation, or `F.interpolate(volume, dim=[-3, -2, -1])` for volumetric data. We can have reasonable defaults if `dim` is not specified (that depends on the shape of the input), to keep backwards compatibility.

The first point will allow us to fuse the `nearest` / `bilinear` / `bicubic` interpolation modes in a single file, while the second point will fuse `temporal` / `spatial` and `volumetric` into the same function.

**Comment by ptrblck** [comment]
Could I try it or is someone already working on it?
CC @colesbury

**Comment by colesbury** [comment]
@SsnL does this overlap with your work?

**Comment by SsnL** [comment]
@colesbury Yes, it would overlap with the grid_sampler optimization I'm doing. However, it would also require porting the upsample/interpolate/resize kernels from THNN to ATen. That is orthogonal with my work. @ptrblck, if you want to, feel free to work on that first, and we can coordinate the later work afterwards.

**Comment by ptrblck** [comment]
Awesome! :)
I'll start porting interpolate to ATen, and get back to you.

**Comment by fmassa** [comment]
@SsnL so you are also modifying the implementation of `interpolate` for the `grid_sample` modifications? Because what I'm proposing here doesn't touch `grid_sample`, only the `*Upsample*` kernels in THNN / THCUNN.

The way I'd see this task would be to avoid porting everything to ATen initially (because it has a lot of code duplication that we would like to avoid), and instead start roughly "from scratch" leveraging the current implementation as a reference.

**Comment by SsnL** [comment]
@fmassa Oh yes, of course you are right. I misread and thought this is also about refactor so `interpolate` methods also share some code with `grid_sample` (which is sill a good eventual goal though). Although, if we are going to "start roughly 'from scratch'" anyways, I'd prefer the new refactor to be rewritten in ATen. What do you think?

**Comment by fmassa** [comment]
Yes, definitely, rewriting it in ATen is what I had in mind

**Comment by fmassa** [comment]
@xmnlab yes, the link went out of date, and you found the right, reference to where I wanted to point to, thanks!

**Comment by ericnakagawa** [comment]
We need to support ATen. See #16158.

**Comment by xmnlab** [comment]
We are working on this issue and we split this into 2 parts: 

1) move from TH to ATen and
2) combine functions that will work for N dimensions.

Initially we are working on an internal branch (https://github.com/Quansight/pytorch/pull/2) before the creation of an official PR.

for step  1) we split that into CPU and GPU. it seems the main structure for CPU is almost done, the building step worked but tests are failing and we are working to fix this issues. after step 1 we can start step 2.

**Comment by fmassa** [comment]
Thanks for the update @xmnlab !

**Comment by xmnlab** [comment]
hey everyone! some updates about this issue: the porting to ATen for cpu code is done https://github.com/pytorch/pytorch/pull/18020

**Comment by ngimel** [comment]
Closing, interpolate is ported to aten, but there's still a lot of code duplication.

**Comment by fmassa** [comment]
Re-opening as code duplication is starting to be fixed in #51653, and follow-up PRs will be applying a similar strategy for other interpolation modes
